quickwit/quickwit-search/src/cluster_client.rs

```diff
@@ -157,6 +170,57 @@ impl ClusterClient {
         // TODO: implement retry
         client.leaf_list_terms(request.clone()).await
     }
+
+    /// Attempts to store a given search context within the cluster.
+    ///
+    /// This function may fail silently, if no clients was available.
+    pub async fn put_kv(&self, key: &[u8], payload: &[u8], ttl: Duration) {
+        let clients = self
+            .search_job_placer
+            .best_nodes_per_affinity(key)
+            .await
+            .take(MAX_PUT_KV_ATTEMPTS);
+
+        // We run the put requests concurrently.
+        // Our target is a replication over TARGET_NUM_REPLICATION nodes, we therefore try to avoid
+        // replicating on more than TARGET_NUM_REPLICATION nodes at the same time. Of
+        // course, this may still result in the replication over more nodes, but this is not
+        // a problem.
+        //
+        // The requests are made in a concurrent manner, up to two at a time. As soon as 2 requests
+        // are successful, we stop.
+        let successful_replication = futures::stream::iter(clients)
+            .map(|mut client| {
+                let put_kv_request = PutKvRequest {
+                    key: key.to_vec(),
+                    payload: payload.to_vec(),
+                    ttl_secs: ttl.as_secs() as u32,
+                };
+                async move { client.put_kv(put_kv_request).await }
+            })
+            .buffer_unordered(TARGET_NUM_REPLICATION)
+            .filter(|put_kv_successful| ready(put_kv_successful.is_ok()))
+            .take(TARGET_NUM_REPLICATION)
+            .count()
+            .await;
+
+        if successful_replication == 0 {
+            error!(successful_replication=%successful_replication,"failed-to-replicate-scroll-context");
+        }
+    }
+
+    /// Returns a search_after context
+    pub async fn get_kv(&self, key: &[u8]) -> Option<Vec<u8>> {
+        let clients = self.search_job_placer.best_nodes_per_affinity(key).await;
+        // On the read side, we attempt to contact up to 6 nodes.
+        for mut client in clients.take(MAX_GET_KV_ATTEMPTS) {
+            let get_request = GetKvRequest { key: key.to_vec() };
+            if let Ok(Some(search_after_resp)) = client.get_kv(get_request.clone()).await {
+                return Some(search_after_resp);
+            }
+        }
```

Comment:
I'd also display a warning here.

Reply:
I am not fond of warn here. It would trigger if you add a node for instance.

Reply:
I made it an `info`